### PR TITLE
Fixed #27554 -- Made prefetched_related aware of RelatedManager's prefetched objects cache

### PR DESCRIPTION
**Pull request by francoisfreitag**
Ticket: https://code.djangoproject.com/ticket/27554

### CODE REVIEW

File: django/db/models/query.py [hunk 1]
```diff
@@ -1418,10 +1418,16 @@ def prefetch_related_objects(model_instances, *related_lookups):
                 # that we can continue with nullable or reverse relations.
                 new_obj_list = []
                 for obj in obj_list:
-                    try:
-                        new_obj = getattr(obj, through_attr)
-                    except exceptions.ObjectDoesNotExist:
-                        continue
+                    if (hasattr(obj, '_prefetched_objects_cache') and through_attr in obj._prefetched_objects_cache):
```
Comment (adamchainz) [inline django/db/models/query.py]: The brackets are unnecessary and should be removed

File: tests/prefetch_related/tests.py [hunk 2]
```diff
@@ -1262,13 +1262,19 @@ def test_bug(self):
         self.assertNotIn(' JOIN ', str(queryset.query))
 
 
-class Ticket25546Tests(TestCase):
+class DirectPrefechedObjectCacheReuseTests(TestCase):
     """
-    Nested prefetch_related() shouldn't trigger duplicate queries for the same
-    lookup.
+    Reusing objects fetched in _prefetched_objects_cache.
```
Comment (adamchainz) [inline tests/prefetch_related/tests.py]: this should be phrased as a "x should do y" sentence, like it was before

File: tests/prefetch_related/tests.py [hunk 3]
```diff
@@ -1358,3 +1368,46 @@ def test_prefetch_with_to_attr(self):
...
+                        'favorite_authors__first_book',
+                    ),
+                ),
+            ),
+        )
+        with self.assertNumQueries(
+            1 +  # AuthorWithAge
+            1 +  # Author
+            1 +  # FavoriteAuthors
+            1    # Book
```
Comment (adamchainz) [inline tests/prefetch_related/tests.py]: not sure of this formatting, could be easier to understand as just `4` with a single comment with the list of models

**Comment by timgraham** [comment]
@charettes, I made a few cosmetic edits, care to review? I'm not sure about the second commit -- sometimes its problematic if tests describe behavior in detail since those comments are unlikely to be updated if the code is refactored in the future.

**Pull request merged**